jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/deploy/WebSocketServerContainerInitializer.java

```diff
@@ -129,54 +132,117 @@ public static boolean isEnabledViaContext(ServletContext context, String keyName
-                    LOG.debug("Dynamic filter add to support JSR356/javax.websocket.server: {}", WebSocketUpgradeFilter.class.getName());
-                WebSocketUpgradeFilter wsuf = WebSocketUpgradeFilter.configureContext(context);
-                context.setAttribute(instanceKey, wsuf);
+                String instanceKey = WebSocketUpgradeFilter.class.getName() + ".SCI";
+                if(context.getAttribute(instanceKey) == null)
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Dynamic filter add to support JSR356/javax.websocket.server: {}", WebSocketUpgradeFilter.class.getName());
+                    WebSocketUpgradeFilter wsuf = WebSocketUpgradeFilter.configureContext(servletContextHandler);
+                    context.setAttribute(instanceKey, wsuf);
+                }
             }
         }
+    }
+
+    /**
+     * Configure the {@link ServletContextHandler} to call {@link WebSocketServerContainerInitializer}
+     * during the {@link ServletContext} initialization phase.
+     *
+     * @param context the context to add listener to
+     */
+    public static void configure(ServletContextHandler context)
+    {
+        context.addEventListener(new SCIOnStartupListener(new WebSocketServerContainerInitializer()));
+    }
+
+    /**
+     * Configure the {@link ServletContextHandler} to call {@link WebSocketServerContainerInitializer}
+     * during the {@link ServletContext} initialization phase.
+     *
+     * @param context the context to add listener to
+     * @param configurator the lambda that is called to allow the {@link ServerContainer} to
+     * be configured during the {@link ServletContext} initialization phase
+     */
+    public static void configure(ServletContextHandler context, BiConsumer<ServletContext, ServerContainer> configurator)
+    {
+        context.addEventListener(new SCIOnStartupListener(new WebSocketServerContainerInitializer())
+        {
+            @Override
+            public void contextInitialized(ServletContextEvent sce)
+            {
+                super.contextInitialized(sce);
+                ServletContext servletContext = sce.getServletContext();
+                ServerContainer serverContainer = (ServerContainer)servletContext.getAttribute(ATTR_JAVAX_SERVER_CONTAINER);
+                configurator.accept(servletContext, serverContainer);
+            }
+        });
+    }
     
-        return jettyContainer;
+    /**
+     * @param context the {@link ServletContextHandler} to use
+     * @return a configured {@link ServerContainer} instance
+     * @throws ServletException if the {@link WebSocketUpgradeFilter} cannot be configured
+     * @deprecated use {@link #configure(ServletContextHandler, BiConsumer)} instead
+     * @see #configure(ServletContextHandler, BiConsumer)
+     */
+    @Deprecated
+    public static ServerContainer configureContext(ServletContextHandler context) throws ServletException
```

Comment:
put this before the two new methods so the diff reads better.

Reply:
Done.